subsys/net/ip/net_context.c

```diff
@@ -332,9 +353,34 @@ static uint16_t find_available_port(struct net_context *context,
 	uint16_t local_port;
 
 	do {
-		local_port = sys_rand16_get() | 0x8000;
+		if (IS_ENABLED(CONFIG_NET_CONTEXT_CLAMP_PORT_RANGE)) {
+			uint16_t upper, lower;
+
+			upper = COND_CODE_1(CONFIG_NET_CONTEXT_CLAMP_PORT_RANGE,
+					    (context->options.port_range >> 16),
+					    (0));
+			lower = COND_CODE_1(CONFIG_NET_CONTEXT_CLAMP_PORT_RANGE,
+					    (context->options.port_range & 0xffff),
+					    (0));
+
+			/* This works the same way as in Linux. If either port
+			 * range is 0, then we use random port. If both are set,
+			 * then we use the range. Also make sure that upper is
+			 * greater than lower.
+			 */
+			if (upper == 0 || lower == 0 || upper < lower) {
+				local_port = sys_rand16_get() | 0x8000;
+			} else {
+				local_port = lower + sys_rand16_get() % (upper - lower);
+
+				NET_DBG("Port range %d - %d, proposing port %d",
+					lower, upper, local_port);
+			}
+		} else {
+			local_port = sys_rand16_get() | 0x8000;
+		}
 	} while (check_used_port(context, NULL, net_context_get_proto(context),
-				 htons(local_port), addr, false, false) == -EEXIST);
+				 htons(local_port), addr, false, false, false) == -EEXIST);
```

Comment:
A bit tricky, but what if someone sets a short range and all of the ports within the range get used - we may spin here indefinitely then? Maybe we should add some retry counter?

Reply:
Good point, I will figure out some fix for that.